File: agent.js
#!/usr/bin/env node

import { readFileSync, writeFileSync, existsSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { randomBytes } from 'crypto';
import AIService from './services/AIService.js';
import { loadTools, getTool } from './tools/toolLoader.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

/**
 * AI Agent for Binance Cryptocurrency Analysis and ML Pipeline Management
 * Now with modular tools supporting Binance price data analysis
 */

class ObscureXAgent {
  constructor(apiKey = null, memoryPath = './agent_memory.json', longTermMemoryPath = './agent_longterm_memory.json') {
    // Initialize AI Service
    this.aiService = new AIService(apiKey);
    
    // Memory management
    this.memoryPath = memoryPath;
    this.longTermMemoryPath = longTermMemoryPath;
    this.memory = this._loadMemory(memoryPath);
    this.longTermMemory = this._loadMemory(longTermMemoryPath);
    
    // Session tracking
    this.sessionId = `session_${Date.now()}_${randomBytes(4).toString('hex')}`;
    
    // Tools will be loaded dynamically
    this.tools = [];
    this._toolsLoaded = false;
  }

  /**
   * Load tools dynamically from the tools directory
   */
  async _ensureToolsLoaded() {
    if (!this._toolsLoaded) {
      this.tools = await loadTools();
      this._toolsLoaded = true;
    }
  }

  /**
   * Load memory from JSON file
   */
  _loadMemory(path) {
    try {
      if (existsSync(path)) {
        const content = readFileSync(path, 'utf-8');
        return JSON.parse(content);
      }
    } catch (error) {
      console.warn(`Failed to load memory from ${path}:`, error.message);
    }
    return { entries: [], metadata: { created: new Date().toISOString() } };
  }

  /**
   * Save memory to JSON file
   */
  _saveMemory(path, memory) {
    try {
      memory.metadata.lastUpdated = new Date().toISOString();
      writeFileSync(path, JSON.stringify(memory, null, 2), 'utf-8');
    } catch (error) {
      console.error(`Failed to save memory to ${path}:`, error.message);
    }
  }

  /**
   * Execute a tool by name with given parameters
   */
  async executeTool(toolName, params) {
    await this._ensureToolsLoaded();
    
    const tool = getTool(this.tools, toolName);
    
    if (!tool) {
      throw new Error(`Tool '${toolName}' not found`);
    }
    
    try {
      // Create context for tool execution
      const context = {
        memory: this.memory,
        longTermMemory: this.longTermMemory,
        sessionId: this.sessionId,
        saveMemory: this._saveMemory.bind(this),
        memoryPath: this.memoryPath,
        longTermMemoryPath: this.longTermMemoryPath,
        aiService: this.aiService
      };
      
      // For tools that need AI, pass aiService in params
      const aiRequiredTools = [
        'generateMLPipeline', 
        'analyzeContext', 
        'getExecutionOptions',
        'recommendOptimizationStrategy',
        'executeAutonomousDecision'
      ];
      
      if (aiRequiredTools.includes(toolName)) {
        params = { ...params, aiService: this.aiService };
      }
      
      const result = await tool.execute(params, context);
      return {
        success: true,
        tool: toolName,
        result: result
      };
    } catch (error) {
      return {
        success: false,
        tool: toolName,
        error: error.message
      };
    }
  }

  /**
   * List all available tools
   */
  async listTools() {
    await this._ensureToolsLoaded();
    
    return this.tools.map(t => ({
      name: t.name,
      description: t.description,
      parameters: t.parameters,
      category: t.category
    }));
  }

  /**
   * Orchestrator: Continuously optimize ML pipeline until MSE threshold is met
   */
  async runOptimizationLoop(config) {
    const {
      dataFile = null,
      initialPrompt = 'Create a price prediction pipeline for Binance data using CatBoost',
      mseThreshold = 0.1,
      maxIterations = 50,
      verbose = true
    } = config;

    console.log('=== Starting Optimization Orchestrator ===');
    console.log(`MSE Threshold: ${mseThreshold}`);
    console.log(`Max Iterations: ${maxIterations}`);
    console.log(`Session ID: ${this.sessionId}\n`);

    // Initialize orchestrator state in memory
    await this.executeTool('storeMemory', {
      key: 'orchestrator_config',
      value: config,
      metadata: { type: 'config' }
    });

    let iteration = 0;
    let currentCode = null;
    let bestMSE = Infinity;
    let bestCode = null;
    let lastMSE = null;

    // Try to load previous best from long-term memory
    const previousBest = await this.executeTool('retrieveLongTermMemory', { key: 'best_pipeline' });
    if (previousBest.success && previousBest.result.success) {
      console.log('Found previous best pipeline in long-term memory');
      console.log(`Previous best MSE: ${previousBest.result.value.mse}\n`);
      if (previousBest.result.value.mse < mseThreshold) {
        console.log('Previous best already meets threshold! Using it as starting point.');
        bestCode = previousBest.result.value.code;
        bestMSE = previousBest.result.value.mse;
      }
    }

    while (iteration < maxIterations) {
      iteration++;
      console.log(`\n--- Iteration ${iteration}/${maxIterations} ---`);

      try {
        // Step 1: Generate or optimize pipeline
        let prompt;
        if (iteration === 1 && !currentCode) {
          prompt = initialPrompt;
          console.log('Generating initial pipeline...');
        } else {
          prompt = await this._generateOptimizationPrompt(currentCode, lastMSE, iteration);
          console.log(`Optimization strategy: ${prompt}`);
        }

        const generateResult = await this.executeTool('generateMLPipeline', {
          existingCode: currentCode,
          prompt: prompt
        });

        if (!generateResult.success) {
          console.error('Failed to generate pipeline:', generateResult.error);
          continue;
        }

        currentCode = generateResult.result.code;
        console.log(`Generated pipeline saved to: ${generateResult.result.filename}`);

        // Step 2: Test the pipeline
        console.log('Testing pipeline...');
        
        let testCode = currentCode;
        if (dataFile) {
          // Replace template comments with actual data file usage
          testCode = testCode.replace(/# df = load_binance_data\([^)]*\)/g, `df = load_binance_data('${dataFile}')`);
          testCode = testCode.replace('# X, y = prepare_features(df)', 'X, y = prepare_features(df)');
          testCode = testCode.replace('# mse, model = train_and_evaluate(X, y)', 'mse, model = train_and_evaluate(X, y)');
          testCode = testCode.replace("# print(f'MSE: {mse}')", "print(f'MSE: {mse}')");
          testCode = testCode.replace('print(\'Pipeline template generated for Binance price data.\')', '');
          testCode = testCode.replace("print('MSE: 0.0')", '');
        }

        const testResult = await this.executeTool('testMLPipeline', {
          pythonCode: testCode
        });

        if (!testResult.success || testResult.result.mse === null) {
          console.error('Failed to test pipeline or extract MSE');
          console.log('Output:', testResult.result?.output || testResult.error);
          
          await this.executeTool('storeMemory', {
            key: `iteration_${iteration}`,
            value: {
              iteration,
              success: false,
              error: testResult.error || 'MSE extraction failed',
              code: currentCode
            },
            metadata: { type: 'iteration_result' }
          });
          continue;
        }

        lastMSE = testResult.result.mse;
        console.log(`MSE: ${lastMSE}`);

        // Step 3: Store iteration result
        await this.executeTool('storeMemory', {
          key: `iteration_${iteration}`,
          value: {
            iteration,
            mse: lastMSE,
            code: currentCode,
            timestamp: new Date().toISOString()
          },
          metadata: { type: 'iteration_result' }
        });

        // Step 4: Update best if improved
        if (lastMSE < bestMSE) {
          bestMSE = lastMSE;
          bestCode = currentCode;
          console.log(`✓ New best MSE: ${bestMSE}`);

          await this.executeTool('storeLongTermMemory', {
            key: 'best_pipeline',
            value: {
              mse: bestMSE,
              code: bestCode,
              iteration,
              timestamp: new Date().toISOString()
            },
            metadata: { type: 'best_result' }
          });

          // Step 5: Check if threshold is met
          if (bestMSE <= mseThreshold) {
            console.log(`\n✓✓✓ SUCCESS! MSE ${bestMSE} meets threshold ${mseThreshold} ✓✓✓`);
            console.log(`Achieved in ${iteration} iterations`);
            
            const finalFile = join(process.cwd(), `best_pipeline_mse_${bestMSE.toFixed(6)}.py`);
            writeFileSync(finalFile, bestCode, 'utf-8');
            console.log(`Best pipeline saved to: ${finalFile}`);
            
            return {
              success: true,
              iterations: iteration,
              bestMSE,
              bestCode,
              filename: finalFile
            };
          }
        } else {
          console.log(`No improvement (best: ${bestMSE})`);
        }

      } catch (error) {
        console.error(`Error in iteration ${iteration}:`, error.message);
        
        await this.executeTool('storeMemory', {
          key: `iteration_${iteration}_error`,
          value: {
            iteration,
            error: error.message,
            stack: error.stack
          },
          metadata: { type: 'error' }
        });
      }

      await new Promise(resolve => setTimeout(resolve, 1000));
    }

    console.log(`\n--- Optimization Complete ---`);
    console.log(`Maximum iterations (${maxIterations}) reached`);
    console.log(`Best MSE achieved: ${bestMSE}`);
    console.log(`Threshold: ${mseThreshold}`);
    
    if (bestCode) {
      const finalFile = join(process.cwd(), `final_pipeline_mse_${bestMSE.toFixed(6)}.py`);
      writeFileSync(finalFile, bestCode, 'utf-8');
      console.log(`Best pipeline saved to: ${finalFile}`);
    }

    return {
      success: bestMSE <= mseThreshold,
      iterations: maxIterations,
      bestMSE,
      bestCode,
      message: bestMSE <= mseThreshold ? 'Threshold met' : 'Threshold not met within max iterations'
    };
  }

  /**
   * Generate optimization prompt using AI or fallback
   */
  async _generateOptimizationPrompt(currentCode, lastMSE, iteration) {
    if (!this.aiService.isAvailable()) {
      const strategies = [
        'Add hyperparameter tuning with GridSearchCV',
        'Add feature engineering with polynomial features',
        'Add cross-validation for better generalization',
        'Try ensemble methods combining multiple models',
        'Add feature selection to remove noise',
        'Increase model complexity',
        'Add regularization to prevent overfitting',
        'Try different preprocessing techniques'
      ];
      return strategies[iteration % strategies.length];
    }

    try {
      const recentResults = await this.executeTool('searchMemory', { query: 'iteration' });
      const history = recentResults.result.results
        .filter(r => r.value.mse !== undefined)
        .slice(-5)
        .map(r => `Iteration ${r.value.iteration}: MSE ${r.value.mse}`)
        .join('\n');

      const prompt = `Based on the ML pipeline optimization history for Binance price data, suggest the next optimization strategy.

Current MSE: ${lastMSE}
Iteration: ${iteration}

Recent history:
${history}

What specific optimization should be applied next to reduce MSE? Be concise and specific.`;

      const systemPrompt = 'You are an expert ML optimization strategist for cryptocurrency price prediction pipelines.';

      return await this.aiService.generateText(prompt, systemPrompt, 200);
    } catch (error) {
      console.warn('Failed to generate AI prompt, using fallback:', error.message);
      return 'Optimize hyperparameters and add cross-validation';
    }
  }

  /**
   * Interactive CLI mode
   */
  async interactiveMode() {
    console.log('ObscureX AI Agent - Interactive Mode');
    console.log('=====================================\n');
    console.log('Available tools:');
    const tools = await this.listTools();
    tools.forEach((tool, idx) => {
      console.log(`${idx + 1}. [${tool.category}] ${tool.name}: ${tool.description}`);
    });
    console.log('\nType "exit" to quit\n');
    console.log('Use agent.executeTool(toolName, params) to execute tools programmatically');
  }
}

// Export for use as a module
export default ObscureXAgent;

// CLI usage
if (import.meta.url === `file://${process.argv[1]}`) {
  const agent = new ObscureXAgent();
  
  if (process.argv.length > 2) {
    const command = process.argv[2];
    
    if (command === 'list-tools') {
      agent.listTools().then(tools => {
        console.log(JSON.stringify(tools, null, 2));
      });
    } else if (command === 'interactive') {
      agent.interactiveMode();
    } else if (command === 'optimize') {
      const dataFile = process.argv[3] || null;
      const threshold = parseFloat(process.argv[4]) || 0.1;
      const maxIter = parseInt(process.argv[5]) || 50;
      
      agent.runOptimizationLoop({
        dataFile,
        mseThreshold: threshold,
        maxIterations: maxIter,
        initialPrompt: 'Create a comprehensive price prediction pipeline for Binance data using CatBoost with proper feature engineering',
        verbose: true
      }).then(result => {
        console.log('\n=== Final Result ===');
        console.log(JSON.stringify(result, null, 2));
      }).catch(error => {
        console.error('Optimization failed:', error);
        process.exit(1);
      });
    } else {
      console.log('Usage:');
      console.log('  node agent.js list-tools                          - List all available tools');
      console.log('  node agent.js interactive                         - Start interactive mode');
      console.log('  node agent.js optimize [dataFile] [threshold] [maxIter] - Run optimization loop');
      console.log('\nExamples:');
      console.log('  node agent.js optimize binance_data.csv 0.05 30   - Optimize until MSE < 0.05');
      console.log('\nOr import the agent in your code:');
      console.log('  import ObscureXAgent from "./agent.js"');
    }
  } else {
    console.log('ObscureX AI Agent - Binance Crypto Analysis\n');
    console.log('Usage:');
    console.log('  node agent.js list-tools                          - List all available tools');
    console.log('  node agent.js interactive                         - Start interactive mode');
    console.log('  node agent.js optimize [dataFile] [threshold] [maxIter] - Run optimization loop');
    console.log('\nExamples:');
    console.log('  node agent.js optimize binance_data.csv 0.05 30   - Optimize until MSE < 0.05');
    console.log('\nOr import the agent in your code:');
    console.log('  import ObscureXAgent from "./agent.js"');
  }
}

File: obscurex/agent.py
#!/usr/bin/env python3
"""
AI Agent for Binance Cryptocurrency Analysis and ML Pipeline Management
Now with modular tools supporting Binance price data analysis
"""

import json
import os
import sys
import asyncio
from datetime import datetime
from pathlib import Path
from typing import Dict, Any, List, Optional
from secrets import token_hex

from .services.ai_service import AIService
from .tools.tool_loader import load_tools, get_tool


class ObscureXAgent:
    """Main AI Agent class for ObscureX."""
    
    def __init__(
        self, 
        api_key: Optional[str] = None,
        memory_path: str = './agent_memory.json',
        long_term_memory_path: str = './agent_longterm_memory.json'
    ):
        """
        Initialize the ObscureX Agent.
        
        Args:
            api_key: Google API key for Gemini (optional, will check env vars)
            memory_path: Path to short-term memory file
            long_term_memory_path: Path to long-term memory file
        """
        # Initialize AI Service
        self.ai_service = AIService(api_key)
        
        # Memory management
        self.memory_path = memory_path
        self.long_term_memory_path = long_term_memory_path
        self.memory = self._load_memory(memory_path)
        self.long_term_memory = self._load_memory(long_term_memory_path)
        
        # Session tracking
        timestamp = int(datetime.now().timestamp() * 1000)
        random_hex = token_hex(4)
        self.session_id = f"session_{timestamp}_{random_hex}"
        
        # Tools will be loaded dynamically
        self.tools: List[Dict[str, Any]] = []
        self._tools_loaded = False
    
    async def _ensure_tools_loaded(self):
        """Load tools dynamically from the tools directory."""
        if not self._tools_loaded:
            self.tools = await load_tools()
            self._tools_loaded = True
    
    def _load_memory(self, path: str) -> Dict[str, Any]:
        """
        Load memory from JSON file.
        
        Args:
            path: Path to memory file
            
        Returns:
            Memory dictionary
        """
        try:
            if os.path.exists(path):
                with open(path, 'r', encoding='utf-8') as f:
                    return json.load(f)
        except Exception as e:
            print(f"Warning: Failed to load memory from {path}: {e}")
        
        return {
            'entries': [],
            'metadata': {'created': datetime.now().isoformat()}
        }
    
    def _save_memory(self, path: str, memory: Dict[str, Any]):
        """
        Save memory to JSON file.
        
        Args:
            path: Path to memory file
            memory: Memory dictionary to save
        """
        try:
            memory['metadata']['lastUpdated'] = datetime.now().isoformat()
            with open(path, 'w', encoding='utf-8') as f:
                json.dump(memory, f, indent=2, ensure_ascii=False)
        except Exception as e:
            print(f"Error: Failed to save memory to {path}: {e}")
    
    async def execute_tool(self, tool_name: str, params: Dict[str, Any]) -> Dict[str, Any]:
        """
        Execute a tool by name with given parameters.
        
        Args:
            tool_name: Name of the tool to execute
            params: Parameters for the tool
            
        Returns:
            Result dictionary with success status and result/error
        """
        await self._ensure_tools_loaded()
        
        tool = get_tool(self.tools, tool_name)
        
        if not tool:
            raise Exception(f"Tool '{tool_name}' not found")
        
        try:
            # Create context for tool execution
            context = {
                'memory': self.memory,
                'longTermMemory': self.long_term_memory,
                'sessionId': self.session_id,
                'saveMemory': self._save_memory,
                'memoryPath': self.memory_path,
                'longTermMemoryPath': self.long_term_memory_path,
                'aiService': self.ai_service
            }
            
            # For tools that need AI, pass aiService in params
            ai_required_tools = [
                'generateMLPipeline',
                'analyzeContext',
                'getExecutionOptions',
                'recommendOptimizationStrategy',
                'executeAutonomousDecision'
            ]
            
            if tool_name in ai_required_tools:
                params = {**params, 'aiService': self.ai_service}
            
            # Execute the tool
            execute_func = tool['execute']
            if asyncio.iscoroutinefunction(execute_func):
                result = await execute_func(params, context)
            else:
                result = execute_func(params, context)
            
            return {
                'success': True,
                'tool': tool_name,
                'result': result
            }
        except Exception as error:
            return {
                'success': False,
                'tool': tool_name,
                'error': str(error)
            }
    
    async def list_tools(self) -> List[Dict[str, Any]]:
        """
        List all available tools.
        
        Returns:
            List of tool metadata dictionaries
        """
        await self._ensure_tools_loaded()
        
        return [
            {
                'name': t.get('name'),
                'description': t.get('description'),
                'parameters': t.get('parameters'),
                'category': t.get('category')
            }
            for t in self.tools
        ]
    
    async def run_optimization_loop(self, config: Dict[str, Any]) -> Dict[str, Any]:
        """
        Orchestrator: Continuously optimize ML pipeline until MSE threshold is met.
        
        Args:
            config: Configuration dictionary with dataFile, mseThreshold, etc.
            
        Returns:
            Results dictionary
        """
        data_file = config.get('dataFile')
        initial_prompt = config.get('initialPrompt', 
                                    'Create a price prediction pipeline for Binance data using CatBoost')
        mse_threshold = config.get('mseThreshold', 0.1)
        max_iterations = config.get('maxIterations', 50)
        verbose = config.get('verbose', True)
        
        if verbose:
            print('=== Starting Optimization Orchestrator ===')
            print(f'MSE Threshold: {mse_threshold}')
            print(f'Max Iterations: {max_iterations}')
            print(f'Session ID: {self.session_id}\n')
        
        # Initialize orchestrator state in memory
        await self.execute_tool('storeMemory', {
            'key': 'orchestrator_config',
            'value': config,
            'metadata': {'type': 'config'}
        })
        
        # Load previous best from long-term memory if available
        best_result = await self.execute_tool('retrieveLongTermMemory', {
            'key': 'best_pipeline'
        })
        
        best_mse = best_result.get('result', {}).get('value', {}).get('mse', float('inf'))
        best_code = best_result.get('result', {}).get('value', {}).get('code')
        
        if verbose and best_code:
            print(f'Loaded previous best MSE: {best_mse}\n')
        
        current_code = best_code
        iteration = 0
        
        while iteration < max_iterations:
            iteration += 1
            
            if verbose:
                print(f'--- Iteration {iteration} ---')
            
            # Generate or optimize ML pipeline
            if current_code:
                prompt = f'Optimize this pipeline to improve MSE. Current MSE: {best_mse}'
                gen_result = await self.execute_tool('generateMLPipeline', {
                    'existingCode': current_code,
                    'prompt': prompt
                })
            else:
                gen_result = await self.execute_tool('generateMLPipeline', {
                    'existingCode': None,
                    'prompt': initial_prompt
                })
            
            if not gen_result.get('success'):
                if verbose:
                    print(f"Error generating pipeline: {gen_result.get('error')}")
                continue
            
            # Test the pipeline
            test_result = await self.execute_tool('testMLPipeline', {
                'pythonCode': gen_result['result'].get('code', '')
            })
            
            if not test_result.get('success'):
                if verbose:
                    print(f"Error testing pipeline: {test_result.get('error')}")
                continue
            
            current_mse = test_result['result'].get('mse', float('inf'))
            
            if verbose:
                print(f'MSE: {current_mse}')
            
            # Update best if improved
            if current_mse < best_mse:
                best_mse = current_mse
                current_code = gen_result['result'].get('code', '')
                
                # Store in long-term memory
                await self.execute_tool('storeLongTermMemory', {
                    'key': 'best_pipeline',
                    'value': {'mse': best_mse, 'code': current_code},
                    'metadata': {'iteration': iteration}
                })
                
                if verbose:
                    print(f'✓ New best MSE: {best_mse}')
            
            # Check if threshold is met
            if best_mse <= mse_threshold:
                if verbose:
                    print(f'\n✓ Threshold met! Best MSE: {best_mse}')
                break
        
        return {
            'bestMSE': best_mse,
            'iterations': iteration,
            'thresholdMet': best_mse <= mse_threshold
        }


def main():
    """Main entry point for CLI usage."""
    import argparse
    
    parser = argparse.ArgumentParser(description='ObscureX AI Agent')
    parser.add_argument('command', nargs='?', default='list-tools',
                       help='Command to run (list-tools, optimize)')
    parser.add_argument('--data-file', help='Data file for optimization')
    parser.add_argument('--threshold', type=float, default=0.1,
                       help='MSE threshold for optimization')
    parser.add_argument('--max-iterations', type=int, default=50,
                       help='Maximum iterations for optimization')
    
    args = parser.parse_args()
    
    # Create agent
    agent = ObscureXAgent()
    
    # Run command
    if args.command == 'list-tools':
        tools = asyncio.run(agent.list_tools())
        print(json.dumps(tools, indent=2))
    elif args.command == 'optimize':
        config = {
            'dataFile': args.data_file,
            'mseThreshold': args.threshold,
            'maxIterations': args.max_iterations,
            'verbose': True
        }
        result = asyncio.run(agent.run_optimization_loop(config))
        print(f"\nOptimization complete:")
        print(f"Best MSE: {result['bestMSE']}")
        print(f"Iterations: {result['iterations']}")
    else:
        print(f"Unknown command: {args.command}")
        sys.exit(1)


if __name__ == '__main__':
    main()
